static/app/utils/formatters.tsx

```diff
@@ -148,70 +148,63 @@ export function getDuration(
 }
 
 /**
- * Returns a human readable exact duration.
- *
+ * Translates seconds into human readable format of seconds, minutes, hours, days, and years
  * e.g. 1 hour 25 minutes 15 seconds
+ *
+ * NOTE: seconds input can be a provided as a decimal
+ * @param  seconds      The number of seconds to be processed
+ * @param  abbreviation abbreviates the suffix
+ * @param  minDuration  pins the response to desired suffix
+ * @return {string}     The phrase describing the amount of time
  */
-export function getExactDuration(seconds: number, abbreviation: boolean = false) {
-  const convertDuration = (secs: number, abbr: boolean): string => {
-    // value in milliseconds
-    const msValue = round(secs * 1000);
-    const value = round(Math.abs(secs * 1000));
-
-    const divideBy = (time: number) => {
-      return {
-        quotient: msValue < 0 ? Math.ceil(msValue / time) : Math.floor(msValue / time),
-        remainder: msValue % time,
-      };
-    };
-
-    if (value >= WEEK) {
-      const {quotient, remainder} = divideBy(WEEK);
-      const suffix = abbr ? t('wk') : ` ${tn('week', 'weeks', quotient)}`;
-
-      return `${quotient}${suffix} ${convertDuration(remainder / 1000, abbr)}`;
-    }
-    if (value >= DAY) {
-      const {quotient, remainder} = divideBy(DAY);
-      const suffix = abbr ? t('d') : ` ${tn('day', 'days', quotient)}`;
-
-      return `${quotient}${suffix} ${convertDuration(remainder / 1000, abbr)}`;
-    }
-    if (value >= HOUR) {
-      const {quotient, remainder} = divideBy(HOUR);
-      const suffix = abbr ? t('hr') : ` ${tn('hour', 'hours', quotient)}`;
-
-      return `${quotient}${suffix} ${convertDuration(remainder / 1000, abbr)}`;
-    }
-    if (value >= MINUTE) {
-      const {quotient, remainder} = divideBy(MINUTE);
-      const suffix = abbr ? t('min') : ` ${tn('minute', 'minutes', quotient)}`;
-
-      return `${quotient}${suffix} ${convertDuration(remainder / 1000, abbr)}`;
+export function getExactDuration(
+  seconds: number,
+  abbreviation: boolean = false,
+  minDuration: string = 'milliseconds'
+) {
+  const operation = seconds < 0 ? Math.ceil : Math.floor;
+  const levels = [
+    [operation(seconds / 604800), abbreviation ? 'wk' : ' weeks'],
+    [operation((seconds % 604800) / 86400), abbreviation ? 'd' : ' days'],
+    [operation(((seconds % 604800) % 86400) / 3600), abbreviation ? 'hr' : ' hours'],
+    [
+      operation((((seconds % 604800) % 86400) % 3600) / 60),
+      abbreviation ? 'min' : ' minutes',
+    ],
+    [
+      operation((((seconds % 604800) % 86400) % 3600) % 60),
+      abbreviation ? 's' : ' seconds',
+    ],
+    [
+      operation((((((seconds * 1000) % 604800000) % 86400000) % 3600000) % 60000) % 1000),
+      abbreviation ? 'ms' : ' milliseconds',
+    ],
+  ];
+  let returntext = '';
+
+  for (let i = 0, max = levels.length; i < max; i++) {
```

Comment:
Have you considered using a greedy approach here? Something like
```
txt = seconds % min duration + min duration unit
value = seconds % min duration

while value % next duration > 1
  txt += value % next duration unit
  value = value % next duration * 60
```

You are also unnecessarily evaluating all the units even though you might break early out of the loop and discard the value entirely